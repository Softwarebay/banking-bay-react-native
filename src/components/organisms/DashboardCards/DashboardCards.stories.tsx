import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { ViewStory } from 'utils/storybook/ViewStory';

import { DashboardCards } from './DashboardCards';

storiesOf('Organisms/DashboardCards', module)
	.addDecorator((getStory) => <ViewStory fullscreen>{getStory()}</ViewStory>)
	.add('Default', () => (
		<DashboardCards
			navText={text('navText', 'Nav Text')}
			onPressCreditCardsIconButton={action('onPressCreditCardsIconButton')}
			onPressProfileIconButton={action('onPressProfileIconButton')}
		/>
	));
