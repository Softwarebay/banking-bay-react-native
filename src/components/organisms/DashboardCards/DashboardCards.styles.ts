import { StyleSheet } from 'react-native';

import { colors } from 'theme';

export const styles = StyleSheet.create({
	bgImage: {
		flex: 1,
	},
	cardContainer: {
		flexDirection: 'row',
	},
	container: {
		backgroundColor: colors.mainDark,
		borderBottomLeftRadius: 30,
		borderBottomRightRadius: 30,
		flex: 0.55,
		overflow: 'hidden',
	},
	divider: {
		marginVertical: 20,
		opacity: 0.3,
	},
	nav: {
		alignItems: 'center',
		justifyContent: 'space-between',
		paddingHorizontal: 20,
		paddingTop: 20,
	},
	navText: {
		color: colors.white,
	},
	paginatorContainer: {
		alignItems: 'center',
		marginTop: -10,
	},
});
