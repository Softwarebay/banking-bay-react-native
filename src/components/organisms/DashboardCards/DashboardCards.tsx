import { useRef, useState } from 'react';
import {
	FlatList,
	Image,
	ImageBackground,
	ImageSourcePropType,
	useWindowDimensions,
	View,
	ViewabilityConfig,
	ViewToken,
} from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';

import { Divider, Flex, IconButton, Paginator, Text } from 'components/atoms';

import { bgDashboard } from 'assets/backgrounds';
import { card1, card2 } from 'assets/images';
import { colors } from 'theme';

import { styles } from './DashboardCards.styles';

interface Props {
	navText: string;
	onPressProfileIconButton: () => void;
	onPressCreditCardsIconButton: () => void;
}

interface Card {
	id: number;
	cardImg: ImageSourcePropType;
}

const cards: Card[] = [
	{
		id: 1,
		cardImg: card1,
	},
	{
		id: 2,
		cardImg: card2,
	},
	{
		id: 3,
		cardImg: card1,
	},
];

export const DashboardCards = ({
	navText,
	onPressProfileIconButton,
	onPressCreditCardsIconButton,
}: Props) => {
	const { width } = useWindowDimensions();
	const [cardIndex, setCardIndex] = useState(0);

	const viewabilityConfig: ViewabilityConfig = {
		viewAreaCoveragePercentThreshold: 50,
	};

	const onViewableItemsChanged = ({
		viewableItems,
	}: {
		viewableItems: ViewToken[];
	}) => {
		const [{ index }] = viewableItems;
		setCardIndex(index ?? 0);
	};

	const viewabilityConfigCallbackPairs = useRef([
		{ viewabilityConfig, onViewableItemsChanged },
	]);

	return (
		<View style={styles.container}>
			<ImageBackground source={bgDashboard} style={[styles.bgImage, { width }]}>
				<SafeAreaView>
					<Flex direction="row" style={styles.nav}>
						<IconButton
							iconColor="#3B5999"
							iconName="profile-icon"
							iconSize={25}
							testID="dashboard-profile-icon-btn-test-id"
							onPress={onPressProfileIconButton}
						/>
						<Text
							style={styles.navText}
							testID="dashboard-nav-text-test-id"
							text={navText}
						/>
						<IconButton
							iconColor={colors.white}
							iconName="credit-cards"
							iconSize={25}
							testID="dashboard-cards-btn-test-id"
							onPress={onPressCreditCardsIconButton}
						/>
					</Flex>
					<Divider />
					<FlatList
						alwaysBounceHorizontal={false}
						contentContainerStyle={styles.cardContainer}
						data={cards}
						decelerationRate="fast"
						initialScrollIndex={cardIndex}
						renderItem={({ item }) => (
							<Image
								source={item.cardImg}
								style={[styles.card, { width: 0.8 * width }]}
							/>
						)}
						showsHorizontalScrollIndicator={false}
						snapToAlignment="center"
						viewabilityConfigCallbackPairs={
							viewabilityConfigCallbackPairs.current
						}
						horizontal
						pagingEnabled
					/>

					<View style={styles.paginatorContainer}>
						<Paginator
							activeColor={colors.white}
							currentIndex={cardIndex}
							slides={cards}
							testID="dashboard-cards-paginator-test-id"
						/>
					</View>
				</SafeAreaView>
			</ImageBackground>
		</View>
	);
};
