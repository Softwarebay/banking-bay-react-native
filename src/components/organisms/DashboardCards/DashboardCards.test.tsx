import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { DashboardCards } from './DashboardCards';

import type { Props } from './DashboardCards';

describe('DashboardCards', () => {
	const mockOnPressCreditCardsIconButton = jest.fn<
		Props['onPressCreditCardsIconButton'],
		[]
	>();
	const mockOnPressProfileIconButton = jest.fn<
		Props['onPressProfileIconButton'],
		[]
	>();
	const props: Props = {
		navText: 'Dashboard Nav',
		onPressCreditCardsIconButton: mockOnPressCreditCardsIconButton,
		onPressProfileIconButton: mockOnPressProfileIconButton,
	};

	it('should match snapshot', () => {
		const tree = testRenderer.create(<DashboardCards {...props} />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render top nav text', () => {
		render(<DashboardCards {...props} />);

		const navText = screen.getByTestId('dashboard-nav-text-test-id');

		expect(navText).toHaveTextContent('Dashboard Nav');
	});

	it('should render profile icon button', () => {
		render(<DashboardCards {...props} />);

		const iconBtn = screen.getByTestId('dashboard-profile-icon-btn-test-id');
		const profileIcon = screen.getByTestId('profile-icon-test-id');

		expect(iconBtn).not.toBeEmptyElement();
		expect(profileIcon).not.toBeNull();
	});

	it('should handle press events on profile icon button', () => {
		render(<DashboardCards {...props} />);

		const iconBtn = screen.getByTestId('dashboard-profile-icon-btn-test-id');

		fireEvent.press(iconBtn);

		expect(mockOnPressProfileIconButton).toHaveBeenCalledTimes(1);
		expect(mockOnPressCreditCardsIconButton).not.toHaveBeenCalled();
	});

	it('should render credit cards icon button', () => {
		render(<DashboardCards {...props} />);

		const iconBtn = screen.getByTestId('dashboard-cards-btn-test-id');
		const cardsIcon = screen.getByTestId('credit-cards-icon-test-id');

		expect(iconBtn).not.toBeEmptyElement();
		expect(cardsIcon).not.toBeNull();
	});
	it('should handle press events on credit cards icon button', () => {
		render(<DashboardCards {...props} />);

		const iconBtn = screen.getByTestId('dashboard-cards-btn-test-id');

		fireEvent.press(iconBtn);

		expect(mockOnPressCreditCardsIconButton).toHaveBeenCalledTimes(1);
		expect(mockOnPressProfileIconButton).not.toHaveBeenCalled();
	});

	it('should render a divider', () => {
		render(<DashboardCards {...props} />);

		const divider = screen.getByTestId('dashboards-cards-divider-test-id');

		expect(divider).not.toBeNull();
	});

	it('should render 3 credit card images', () => {
		render(<DashboardCards {...props} />);

		const cardImgs = screen.getAllByTestId(/(card1|card2)-img-test-id/);

		expect(cardImgs).toHaveLength(3);
	});

	it('should render paginator', () => {
		render(<DashboardCards {...props} />);

		const paginator = screen.getByTestId('dashboard-cards-paginator-test-id');

		expect(paginator).not.toBeNull();
	});
});
