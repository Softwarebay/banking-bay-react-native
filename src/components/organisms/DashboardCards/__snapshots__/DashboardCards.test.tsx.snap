// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardCards should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#1B1D4D",
      "borderBottomLeftRadius": 30,
      "borderBottomRightRadius": 30,
      "flex": 0.55,
      "overflow": "hidden",
    }
  }
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "width": 750,
        },
      ]
    }
  >
    <Image
      source={1}
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": undefined,
            "width": 750,
          },
          undefined,
        ]
      }
    />
    <RNCSafeAreaView
      edges={
        Array [
          "bottom",
          "left",
          "right",
          "top",
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
            },
            Object {
              "flexDirection": "row",
            },
            Object {
              "alignItems": "center",
              "justifyContent": "space-between",
              "paddingHorizontal": 20,
              "paddingTop": 20,
            },
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          hitSlop={20}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              undefined,
              false,
            ]
          }
          testID="dashboard-profile-icon-btn-test-id"
        >
          <SvgMock
            color="#3B5999"
            height={25}
            testID="profile-icon-test-id"
            width={25}
          />
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#4C4C60",
                "fontFamily": "MulishRegular",
                "fontSize": 17,
                "lineHeight": 27.2,
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
          testID="dashboard-nav-text-test-id"
        >
          Dashboard Nav
        </Text>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          hitSlop={20}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              undefined,
              false,
            ]
          }
          testID="dashboard-cards-btn-test-id"
        >
          <SvgMock
            color="#FFFFFF"
            height={25}
            testID="credit-cards-icon-test-id"
            width={25}
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "borderColor": "#CED6E1",
              "borderTopWidth": 1,
            },
            Object {
              "marginVertical": 20,
              "opacity": 0.3,
            },
          ]
        }
        testID="dashboards-cards-divider-test-id"
      />
      <RCTScrollView
        contentContainerStyle={
          Object {
            "flexDirection": "row",
          }
        }
        data={
          Array [
            Object {
              "id": 1,
              "img": "card1",
            },
            Object {
              "id": 2,
              "img": "card2",
            },
            Object {
              "id": 3,
              "img": "card1",
            },
          ]
        }
        decelerationRate="fast"
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={true}
        initialScrollIndex={0}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        pagingEnabled={true}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        showsHorizontalScrollIndicator={false}
        snapToAlignment="center"
        stickyHeaderIndices={Array []}
        viewabilityConfigCallbackPairs={
          Array [
            Object {
              "onViewableItemsChanged": [Function],
              "viewabilityConfig": Object {
                "viewAreaCoveragePercentThreshold": 50,
              },
            },
          ]
        }
      >
        <View>
          <View
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <Image
              source={1}
              style={
                Array [
                  Object {
                    "height": undefined,
                  },
                  Object {
                    "aspectRatio": 1.5472312703583062,
                  },
                  Object {
                    "width": 750,
                  },
                ]
              }
              testID="card1-img-test-id"
            />
          </View>
          <View
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <Image
              source={1}
              style={
                Array [
                  Object {
                    "height": undefined,
                  },
                  Object {
                    "aspectRatio": 1.5472312703583062,
                  },
                  Object {
                    "width": 750,
                  },
                ]
              }
              testID="card2-img-test-id"
            />
          </View>
          <View
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <Image
              source={1}
              style={
                Array [
                  Object {
                    "height": undefined,
                  },
                  Object {
                    "aspectRatio": 1.5472312703583062,
                  },
                  Object {
                    "width": 750,
                  },
                ]
              }
              testID="card1-img-test-id"
            />
          </View>
        </View>
      </RCTScrollView>
      <View
        style={
          Object {
            "alignItems": "center",
            "marginTop": -10,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
              },
              Object {
                "flexDirection": "row",
              },
              undefined,
            ]
          }
          testID="dashboard-cards-paginator-test-id"
        >
          <View
            style={
              Array [
                Object {
                  "borderColor": "#1B1D4D",
                  "borderRadius": 100,
                  "borderWidth": 3,
                  "height": 8,
                  "marginHorizontal": 8,
                  "opacity": 0.2,
                  "width": 8,
                },
                Object {
                  "opacity": 1,
                },
                Object {
                  "borderColor": "#FFFFFF",
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "borderColor": "#1B1D4D",
                  "borderRadius": 100,
                  "borderWidth": 3,
                  "height": 8,
                  "marginHorizontal": 8,
                  "opacity": 0.2,
                  "width": 8,
                },
                false,
                Object {
                  "borderColor": "#FFFFFF",
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "borderColor": "#1B1D4D",
                  "borderRadius": 100,
                  "borderWidth": 3,
                  "height": 8,
                  "marginHorizontal": 8,
                  "opacity": 0.2,
                  "width": 8,
                },
                false,
                Object {
                  "borderColor": "#FFFFFF",
                },
              ]
            }
          />
        </View>
      </View>
    </RNCSafeAreaView>
  </View>
</View>
`;
